core/json/json_path.rs

```diff
@@ -0,0 +1,137 @@
+use pest::Parser as P;
+use pest_derive::Parser;
+
+#[derive(Parser)]
+#[grammar = "json/json.pest"]
+#[grammar = "json/json_path.pest"]
+struct Parser;
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum PathElement {
+    Root(),
+    Key(String),
+    ArrayLocator(i32),
+}
+
+#[derive(Debug)]
+pub struct JsonPath {
+    pub elements: Vec<PathElement>,
+}
+
+// Parses path into a Vec of Strings, where each string is a key or an array locator.
+pub fn json_path(path: &str) -> crate::Result<JsonPath> {
+    let parsed = Parser::parse(Rule::path, path);
+
+    if let Ok(mut parsed) = parsed {
+        let mut result = vec![];
+        let parsed = parsed.next().unwrap();
+        for pair in parsed.into_inner() {
+            match pair.as_rule() {
+                Rule::EOI => (),
+                Rule::root => result.push(PathElement::Root()),
+                Rule::json_path_key => result.push(PathElement::Key(pair.as_str().to_string())),
+                Rule::array_locator => {
+                    let mut array_locator = pair.into_inner();
+                    let index_or_negative_indicator = array_locator.next().unwrap();
+
+                    match index_or_negative_indicator.as_rule() {
+                        Rule::negative_index_indicator => {
+                            let negative_offset = array_locator.next().unwrap();
+                            // TODO: sqlite is able to parse arbitrarily big numbers, but they
+                            //  always get overflown and cast to i32. Handle this.
+                            let parsed = negative_offset
+                                .as_str()
+                                .parse::<i128>()
+                                .unwrap_or(i128::MAX);
+
+                            result.push(PathElement::ArrayLocator(-parsed as i32));
+                        }
+                        Rule::array_offset => {
+                            let array_offset = index_or_negative_indicator.as_str();
+                            // TODO: sqlite is able to parse arbitrarily big numbers, but they
+                            //  always get overflown and cast to i32. Handle this.
+                            let parsed = array_offset.parse::<i128>().unwrap_or(i128::MAX);
+
+                            result.push(PathElement::ArrayLocator(parsed as i32));
+                        }
+                        _ => unreachable!(
+                            "Unexpected rule: {:?}",
+                            index_or_negative_indicator.as_rule()
+                        ),
+                    }
+                }
+                _ => {
```

Comment:
can this branch be hit? doesn't the parser produce discreet unions where we can handle every possible union variant?